coderd/autobuild/executor/lifecycle_executor.go

```diff
@@ -89,77 +90,104 @@ func (e *Executor) runOnce(t time.Time) Stats {
 		stats.Error = err
 	}()
 	currentTick := t.Truncate(time.Minute)
-	err = e.db.InTx(func(db database.Store) error {
-		// TTL is set at the workspace level, and deadline at the workspace build level.
-		// When a workspace build is created, its deadline initially starts at zero.
-		// When provisionerd successfully completes a provision job, the deadline is
-		// set to now + TTL if the associated workspace has a TTL set. This deadline
-		// is what we compare against when performing autostop operations, rounded down
-		// to the minute.
-		//
-		// NOTE: If a workspace build is created with a given TTL and then the user either
-		//       changes or unsets the TTL, the deadline for the workspace build will not
-		//       have changed. This behavior is as expected per #2229.
-		eligibleWorkspaces, err := db.GetWorkspacesAutostart(e.ctx)
-		if err != nil {
-			return xerrors.Errorf("get eligible workspaces for autostart or autostop: %w", err)
-		}
 
-		for _, ws := range eligibleWorkspaces {
-			// Determine the workspace state based on its latest build.
-			priorHistory, err := db.GetLatestWorkspaceBuildByWorkspaceID(e.ctx, ws.ID)
-			if err != nil {
-				e.log.Warn(e.ctx, "get latest workspace build",
-					slog.F("workspace_id", ws.ID),
-					slog.Error(err),
-				)
-				continue
-			}
+	// TTL is set at the workspace level, and deadline at the workspace build level.
+	// When a workspace build is created, its deadline initially starts at zero.
+	// When provisionerd successfully completes a provision job, the deadline is
+	// set to now + TTL if the associated workspace has a TTL set. This deadline
+	// is what we compare against when performing autostop operations, rounded down
+	// to the minute.
+	//
+	// NOTE: If a workspace build is created with a given TTL and then the user either
+	//       changes or unsets the TTL, the deadline for the workspace build will not
+	//       have changed. This behavior is as expected per #2229.
+	eligibleWorkspaces, err := e.db.GetWorkspacesAutostart(e.ctx)
```

Comment:
Since we're going to be re-fetching the workspaces in their own Tx anyway, how about we just return the IDs of autostart-eligible workspaces now?